README.md

```diff
@@ -1,125 +1,167 @@
-    { 'shellcheck', auto_update = true },
-
-    'stylua',
-    'shellcheck',
-    'editorconfig-checker',
-    'revive',
-    'shellcheck',
-    'shfmt',
-    'staticcheck',
-    'vint',
-  },
-
-	null_ls_sources = {
-		-- Will install the all necessary tools from Mason, in this case `write-good`.
-		-- Source must map to the source name in `null-ls`.
-		-- Pinning a version or toggling auto_update per tool is not supported.
-		'write_good',
-	},
-
-  -- if set to true this will check each tool for updates. If updates
-  -- are available the tool will be updated. This setting does not
-  -- affect :MasonNullLsUpdate or :MasonNullLsInstall.
-  -- Default: false
-  auto_update = false,
-
-  -- if set to true, the tool will check null-ls for the available sources,
-  -- and then install the dependency(ies) if available in mason.
-  -- Default: true
-  automatic_installation = true
-}
+require("mason").setup()
+require("mason-null-ls").setup()
 ```
 
-## Commands
+Refer to the [Configuration](#configuration) section for information about which settings are available.
+
+
+# Commands
+
+-   `:NullInstall [<source>...]` - installs the provided sources
+-   `:NullUninstall <source> ...` - uninstalls the provided sources
+
 
-`:MasonNullLsInstall` - only installs tools that are missing or at the incorrect version
-`:MasonNullLsUpdate` - install missing tools and update already installed tools
+# Configuration
 
-## Events
+You may optionally configure certain behavior of `mason-null-ls.nvim` when calling the `.setup()` function. Refer to
+the [default configuration](#default-configuration) for a list of all available settings.
 
-Upon completion of any `mason-null-ls` initiated installation/update a user event will be
-emitted named `MasonNullLsUpdateCompleted`. To use this event you can setup an event handler like so:
+Example:
 
 ```lua
-  vim.api.nvim_create_autocmd('User', {
-    pattern = 'MasonNullLsUpdateCompleted',
-    callback = function()
-      vim.schedule(print 'mason-null-ls has finished')
-    end,
-  })
+require("mason-null-ls").setup({
+    ensure_installed = { "stylua", "jq" }
+})
 ```
 
-## Suggestions / Complaints / Help
+## Default configuration
+
+```lua
+local DEFAULT_SETTINGS = {
+    -- A list of sources to automatically install if they're not already installed. Example: { "rust_analyzer@nightly", "sumneko_lua" }
+    -- This setting has no relation with the `automatic_installation` setting.
+    ensure_installed = {},
+}
+```
+
+
+# Setup handlers usage
 
-Create an issue at: https://github.com/jayp0521/mason-null-ls.nvim/issues
+The `setup_handlers()` function provides a dynamic way of setting up sources and any other needed logic, It can also do that during runtime.
+
+```lua
+  local null_ls = require 'null-ls'
+
+  require 'mason-null-ls'.setup { ensure_installed = {'stylua', 'jq'} }
+  
+  require 'mason-null-ls'.setup_handlers {
+    function(source_name)
+      -- all sources with no handler get passed here
```

Comment:
I wonder if this snippet should recommend the following:

```suggestion
      for _, source in ipairs(null_ls.get_source({ name = source_name })) do
          null_ls.register(source)
      end
```

Reply:
Hm nvm, I had missed that `null_ls.get_source()` only returns already registered sources. I think the general idea is clear though. I think this `.setup_handlers()` function would be pretty awkward without some way to access all underlying sources

Reply:
Hmm, I don't think `null-ls` provides a way to query builtin sources.
The author might be future proofing in anticipation for a source with multiple different methods or they just didn't like integrating the builtin sources further into the plugin.

Either way your suggestion makes a lot of sense (and is mentioned in the PR message within the code snippet as a comment), the only way I can think of implementing it is by passing `source_method` into the default handler so we can do
```lua
require ('mason-null-ls').setup_handlers {
  function(source_name, source_method) {
    null_ls.register(null_ls.builtins[source_method][source_name])
  }
}
```

lmk what you think

Reply:
I think this could probably be automatically generated, I'd be happy to port over the script capabilities + Github workflows. I'm also not sure there's a similar need for null-ls so the API could probably be simplified a bit. On mobile atm but will provide an alternative approach later